platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -33,66 +33,76 @@ These methods correspond to the state lifecycle as follows:
+- Keep track of the references to the latest immutable state and the latest mutable state.
+- Restrict mutability to a single state. That is, only one state object should be mutable at any given time.
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+    void setInitialState(State initialState);
+
+    State getLatestImmutableState();
 
-Additionally, there should be a configuration parameter that limits the number of states kept in memory. When the limit is reached, the oldest state should be evicted.
-Currently, we rely on the implicit garbage collection. We have a reference count mechanism to achieve this.
-We want to get rid of this mechanism and rely on the explicit eviction mechanism.
+    State getMutableState();
+}
+```
+
+**Imporatant**: It's up to the client code to make sure that the state is released once it's no longer needed.
+Without this, the state will not be evicted from memory and eventually the application will run out of memory.
 
 The following diagram illustrates an example of the states in memory throughout their lifecycles:
 
-[![Multistate management](multi-state-management.png)](multi-state-management.png)
+[![Multistate management](multi-states.svg)](multi-states.svg)
+
+Also, `swirlds-state-api` needs a `SnapshotManager` interface with an implementation `SnapshotManagerImpl` in `swirlds-state-impl` that will have the following responsibilities:
+- Load snapshots from the disk. A state loaded from a snapshot must be immutable and available for look up by the round number.
+- Evict the loaded state by the round number when the state is no longer needed.
+
+This functionality is separated from the `StateLifecycleManager` for the following reasons:
+- a state loaded from a snapshot is _always_ immutable, and therefore there is no notion of mutability here.
+- the mechanism of maintaining the loaded states in memory is different. This functionality doesn't rely on the reference counting.
+Instead, it has explicit methods to load a state from a snapshot, and evict the state by the round number when it's no longer needed.
 
 The interface will look something like this:
 
 ```java
-public interface StateManager {
-    State createEmptyState();
-    
-    State getLatestImmutableState();
-    
-    State getMutableState();
-    
-    State getState(long round);
-    
-    void evictState(long round);
-    
+public interface SnapshotManager {
+
+    State getLoadedState(long round);
+
     State loadSnapshot(Path sourcePath);
+
+    void evictLoadedState(long round);
 }
 ```
 
-Note that the loaded snapshot should be treated as one of the immutable states and should be available via `getState`.
-
 ### Metrics
 
-We need to monitor the number of states that are kept in memory to ensure there is no memory leak.
+We need to monitor the number of states that are kept in memory by `SnapshotManager` to ensure there is no memory leak.
```

Comment:
I think this can be removed

Reply:
`StateLifecycleManager` will maintain a list of states previously loaded with `loadSnapshot` method. We need to monitor the size of this list to make sure that we don't have unused state accumulating in memory or they may be all in use but it's just too many of them which may cause OOM. 

Alternatively, we can get rid of `State getLoadedState(long round)` method and use `loadSnapshot` whenever we need to access a state on the disk.